Allow the construction of self-referential `Ref`s

garyb opened the pull request.

natefaubion commented on src/Effect/Ref.js:
```diff
@@ -1,8 +1,10 @@
   return function () {
-    return { value: val };
+    var ref = { value: null };
+    ref.value = f(ref);
```
I think it would be safer to throw an exception if the value isn't set yet, otherwise you break type-safety since `null` can silently sneak into your program.

garyb replied:
How would we do that?

natefaubion replied:
You have to guard read/modify, unfortunately.
```js
var EMPTY = {};

exports.self = function (f) {
  return function () {
    var ref = { value: EMPTY };
    ref.value = f(ref);
    return ref;
  };
};

exports.read = function (ref) {
  return function () {
    if (ref.value === EMPTY) {
      throw new Error("Uninitialized ref");
    }
    return ref.value;
  };
};
```

I think this is a reason why I'd rather keep it out of the lib. If a user wants to `unsafeCoerce unit` a value, that's their choice, but I don't want to expose an API that is dangerously and silently unsafe.

garyb replied:
Can you concoct a scenario in which this thing is unsafe, outside of using more `unsafeCoerce` and such? The `Ref s -> s` function is pure, so it has no opportunity to do anything with the new `Ref` other than use it as a value.

natefaubion replied:
The other alternative is to do the `monad-fix` thing of passing in `Unit -> Ref a` instead. By passing in the Ref directly, you have to guard every read, but if you pass in a thunk, you only have to do the guard in the thunk.

natefaubion replied:
Oh, I see yes. If the callback is pure, I guess it would be OK. I was thinking it was in Effect.

natefaubion commented on src/Effect/Ref.purs:
```diff
@@ -13,7 +13,12 @@ import Effect (Effect)
 -- | Create a new mutable reference containing the specified value.
-foreign import new :: forall s. s -> Effect (Ref s)
+new :: forall s. s -> Effect (Ref s)
+new s = self (const s)
```
I don't really like that `new` requires an unnecessary closure allocation with `const` on top of the ref allocation and then Effect thunk allocation.

garyb replied:
Yeah, I guess the whole thing is FFI already so one more function won't do any harm.

paldepind commented on src/Effect/Ref.purs:
```diff
@@ -15,6 +15,10 @@ foreign import data Ref :: Type -> Type
 
+-- | Create a new mutable reference containing a value that can refer to the
+-- | `Ref` being created.
+foreign import self :: forall s. (Ref s -> s) -> Effect (Ref s)
```
What about naming this function `newSelf` instead to highlight that it is a variant of the `new` function? To me, the name `self` and nothing else seems a bit vague. If I saw the name `newSelf` I would immediately know that the function is related to `new`.

JordanMartinez commented:
What's the status of this PR?

garyb commented:
Is "forgotten entirely about" a valid status? :smile:

It sounds like the review concern was addressed, but there's no explicit approval.

hdgarrood commented:
This looks good to me, although I think I’d prefer the function was called `newSelf` too.

JordanMartinez commented:
> newSelf

Same. @garyb How quickly would you be able to update this PR? Or should someone else submit a PR to your branch to get the name changed?

garyb commented:
Done! I wonder if there's still a better name though, as `newSelf` is a bit weird. Something like `newKnot` perhaps.

JordanMartinez commented:
> I wonder if there's still a better name though, as newSelf is a bit weird. Something like knot or newKnot perhaps.

Why? What about `knot` or `newKnot` makes you think it's better?

JordanMartinez commented:
Looks like CI is failing because the tests weren't updated:
```
* Building project in /home/travis/build/purescript/purescript-refs

[1/1 UnknownName] test/Main.purs:39:10

  39    ref <- Ref.self \ref -> RefBox { ref, value: 0 }

               ^^^^^^^^

  

  Unknown value Ref.self
```

hdgarrood commented:
People use the phrase “tying the knot” to mean constructing a structure which references itself: https://wiki.haskell.org/Tying_the_Knot

JordanMartinez commented:
Let's assume you come to a language and codebase you've never seen before and you only see this: `ref.newSelf` and later `ref.newKnot`. 
By just looking at those isolated instances, what immediately comes to mind?

IMHO, I think the `newSelf` makes me think more of what this PR is doing. The `newKnot` would require me to know about the "Tying the Knot" phrase and I would need to connect that concept with this usage rather than assume that `knot` is some domain concept and I'm creating a new one.

natefaubion commented:
newWithSelf?

JordanMartinez commented:
Sure.

garyb commented:
Renamed again, and updated the tests this time!

@JordanMartinez re: people coming to the library looking for this function, realistically I'm not sure how often that will happen. I needed to do this in Halogen, and used an `unsafeCoerce` hack to make it work, but outside of that case I can't think of another occasion I've needed it. In Halogen it's in a situation where something is stored in an existential form, but internally it needs to be able to do things with its full set of type parameters, so... pretty niche. 

Maybe there are some other uses cases for it though, I dunno. I've definitely used self referential `AVar`s more often. But they don't have this problem since they can be created as empty-but-not-`Maybe` due to their nature, unlike `Ref`.

I think providing a way to do it in the library rather than having to do an ad-hoc `unsafeCoerce` hack for the rare cases it's needed is still worthwhile, even if it is rare, since it's not possible to misuse this the way I theoretically could in Halogen.
